tools/cli/app_test.go

```diff
@@ -0,0 +1,388 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package cli
+
+import (
+	"testing"
+	"time"
+
+	"github.com/golang/mock/gomock"
+	"github.com/pborman/uuid"
+	"github.com/stretchr/testify/suite"
+	"github.com/uber/cadence/common"
+	"github.com/urfave/cli"
+
+	"go.uber.org/cadence/.gen/go/cadence/workflowserviceclient"
+	"go.uber.org/cadence/.gen/go/cadence/workflowservicetest"
+	"go.uber.org/cadence/.gen/go/shared"
+)
+
+type cliAppSuite struct {
+	suite.Suite
+	app     *cli.App
+	service *workflowservicetest.MockClient
+}
+
+type workflowClientBuilderMock struct {
+	service workflowserviceclient.Interface
+}
+
+func (mock *workflowClientBuilderMock) BuildServiceClient(c *cli.Context) (workflowserviceclient.Interface, error) {
+	return mock.service, nil
+}
+
+// this is the mock for yarpcCallOptions, make sure length are the same
+var callOptions = []interface{}{gomock.Any(), gomock.Any(), gomock.Any()}
+
+var commands = []string{
+	"domain", "d",
+	"workflow", "wf",
+	"tasklist", "tl",
+}
+
+var domainName = "cli-test-domain"
+
+func TestCLIAppSuite(t *testing.T) {
+	s := new(cliAppSuite)
+	suite.Run(t, s)
+}
+
+func (s *cliAppSuite) SetupSuite() {
+	s.app = NewCliApp()
+}
+
+func (s *cliAppSuite) SetupTest() {
+	mockCtrl := gomock.NewController(s.T())
+	s.service = workflowservicetest.NewMockClient(mockCtrl)
+	SetBuilder(&workflowClientBuilderMock{service: s.service})
+}
+
+func (s *cliAppSuite) TestAppCommands() {
+	for _, test := range commands {
+		cmd := s.app.Command(test)
+		s.NotNil(cmd)
+	}
+}
+
+func (s *cliAppSuite) TestDomainRegister() {
+	s.service.EXPECT().RegisterDomain(gomock.Any(), gomock.Any(), callOptions...).Return(nil)
+	err := s.app.Run([]string{"", "--do", domainName, "domain", "register"})
+	s.Nil(err)
+}
+
+func (s *cliAppSuite) TestDomainRegister_DomainExist() {
+	s.service.EXPECT().RegisterDomain(gomock.Any(), gomock.Any(), callOptions...).Return(&shared.DomainAlreadyExistsError{})
+	err := s.app.Run([]string{"", "--do", domainName, "domain", "register"})
+	s.Nil(err)
```

Comment:
this does not test anything, because func RegisterDomain(c *cli.Context) return nothing so it always will be nil.

Reply:
Understand, currently these unit tests do not test the "real output", will probably revisit in future.